### PR TITLE
Input validation with error translations

### PR DESCRIPTION
**Pull request by ao-armen**

### CODE REVIEW

File: src/containers/App/index.js [hunk 1]
```diff
@@ -1,21 +1,51 @@
...
-  const { t } = useTranslation();
+  const [t] = useTranslation();
+  const { register, handleSubmit, errors, triggerValidation } = useForm({
+    validationSchema: appSchema(t),
+    mode: 'onBlur',
+  });
+
+  useEffect(() => {
+    return () => errors && triggerValidation();
+  }, [t]);
```
Comment (hpilosyan) [inline src/containers/App/index.js]: What this does [t]?

File: src/containers/App/index.js [hunk 2]
```diff
@@ -1,21 +1,51 @@
...
+    <form onSubmit={handleSubmit(onSubmit)} className={styles.app}>
       <header className={styles.app_header}>
         <img src={logo} className={styles.app_logo} alt="logo" />
         <p>{t('login.username')}</p>
         <LanguageContainer />
+
+        <br />
+
+        <Validation name="firstName" showMessage={true} errors={errors}>
+          <input className={styles.lol} name="firstName" type="text" ref={register}></input>
```
Comment (hpilosyan) [inline src/containers/App/index.js]: styles.lol ?

File: src/containers/App/index.js [hunk 3]
```diff
@@ -1,21 +1,51 @@
...
 function App() {
-  const { t } = useTranslation();
+  const [t] = useTranslation();
+  const { register, handleSubmit, errors, triggerValidation } = useForm({
+    validationSchema: appSchema(t),
+    mode: 'onBlur',
+  });
+
+  useEffect(() => {
+    return () => errors && triggerValidation();
```
Comment (hpilosyan) [inline src/containers/App/index.js]: Why is this needed?

File: src/validations/app.js [hunk 4]
```diff
@@ -0,0 +1,10 @@
+import * as Yup from 'yup';
+
+const appSchema = (t) => Yup.object().shape({
```
Comment (hpilosyan) [inline src/validations/app.js]: Any other way to access t here?

File: src/components/Validation/index.js [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+    <>
+      {React.Children.map(children, child => {
+        return React.cloneElement(child, {
+          className: classnames(child.props.className, {
+            [styles.invalid]: errors[name],
+          }),
+          name: name,
+        });
+      })}
+      {errors[name] && showMessage && (
```
Comment (hpilosyan) [inline src/components/Validation/index.js]: Now you handle client side errors. Did you consider server side errors?

File: src/containers/App/index.js (hunk 1 above)
Comment (ao-armen) [reply src/containers/App/index.js]: useEffect takes as argument and array of values, every time one of those values changes useEffect will work.

File: src/containers/App/index.js (hunk 3 above)
Comment (ao-armen) [reply src/containers/App/index.js]: if I write just errors && triggerValidation() inputs would be validated when component mounted, when returning function like that it validates on blur out

File: src/validations/app.js (hunk 4 above)
Comment (ao-armen) [reply src/validations/app.js]: I can not use useTranslate() here, it's a hook so can be called inside React function component or custom React hook function
Or I can write validations right in validationSchema  and not import it

Comment (hpilosyan) [reply src/validations/app.js]: We were talking about using i18n instance directly

File: src/i18n/translations/hy.json [hunk 6]
```diff
@@ -1,4 +1,5 @@
 {
+  "is_required": "Այս դաշտը պարտադիր է",
```
Comment (hpilosyan) [inline src/i18n/translations/hy.json]: We should probably have validation section, which contains "required"

File: src/i18n/translations/hy.json [hunk 7]
```diff
@@ -10,5 +11,8 @@
   },
   "home": {
     "title": "Գլխավոր"
+  },
+  "error_codes": {
+    "repeated_first_name": "խնդրում ենք ընտրել այլ անուն"
```
Comment (hpilosyan) [inline src/i18n/translations/hy.json]: What's this one and how is it different from is_required?

File: src/components/LoginScreen/index.js [hunk 8]
```diff
@@ -1,37 +1,65 @@
...
+    );
   };
 
   return (
     <div id={styles.login_screen}>
       <div className={styles.login_body}>
-        <form onSubmit={handleFormSubmit}>
+        <form onSubmit={handleSubmit(onSubmit)}>
+
+          <LanguageContainer />
```
Comment (hpilosyan) [inline src/components/LoginScreen/index.js]: Why do we need this here?

File: src/components/LoginScreen/index.js [hunk 9]
```diff
@@ -1,37 +1,65 @@
-import React from "react";
+import React, { useEffect } from 'react';
+import { useTranslation } from 'react-i18next';
+import appSchema from '../../validations/app';
```
Comment (hpilosyan) [inline src/components/LoginScreen/index.js]: Lets clear this up and import libraries first and then local modules

File: src/components/LoginScreen/index.js [hunk 10]
```diff
@@ -1,37 +1,65 @@
...
+  });
+
+  useEffect(() => {
+    return () => errors && triggerValidation();
+  }, [t]);
 
-  const handleFormSubmit = event => {
-    event.preventDefault();
+  const onSubmit = data => {
+    setError(
```
Comment (hpilosyan) [inline src/components/LoginScreen/index.js]: Hmmmm, do we need to manually do this every time?

File: src/components/LoginScreen/index.js [hunk 11]
```diff
@@ -1,37 +1,65 @@
...
 
 const LoginScreen = props => {
   const { t } = useTranslation();
+  const { register, handleSubmit, errors, triggerValidation, setError } = useForm({
+    validationSchema: appSchema(t),
+    mode: 'onBlur',
+  });
+
+  useEffect(() => {
+    return () => errors && triggerValidation();
```
Comment (hpilosyan) [inline src/components/LoginScreen/index.js]: I don't understand what you do here

File: src/components/LoginScreen/index.js [hunk 12]
```diff
@@ -1,37 +1,65 @@
...
+  }, [t]);
 
-  const handleFormSubmit = event => {
-    event.preventDefault();
+  const onSubmit = data => {
+    setError(
+      "username",
```
Comment (hpilosyan) [inline src/components/LoginScreen/index.js]: You do something wrong. This basic example doesn't contain all the manual things you do:
https://github.com/react-hook-form/react-hook-form/blob/master/examples/basicValidation.tsx

File: src/components/LoginScreen/index.js (hunk 10 above)
Comment (ao-armen) [reply src/components/LoginScreen/index.js]: No, I just left an example here in case of if there be some values that are valid but shouldn't be in request, like if we allow phone number input to be without +374 but we want a to show error/warning without doing a request

File: src/components/LoginScreen/index.js (hunk 11 above)
Comment (ao-armen) [reply src/components/LoginScreen/index.js]: Returning function in useEffect works like componentWillUnmount, now useEffect is listening to "t" so function will run at the end of change, it wont validate on mount and will validate only after change of "t"

File: src/components/LoginScreen/index.js (hunk 10 above)
Comment (hpilosyan) [reply src/components/LoginScreen/index.js]: WE need clean examples

**Comment by ao-tigran** [comment]
@ao-armen have you fixed comments?

**Comment by ao-armen** [comment]
@ao-tigran No, not yet.

**Comment by ao-tigran** [comment]
@ao-armen whenever there are comments please concentrate on fixing them and then only work only continue working on new task.

**Comment by ao-tigran** [comment]
@ao-armen have you rebased this?

File: src/validations/app.js [hunk 13]
```diff
@@ -0,0 +1,11 @@
+import * as Yup from 'yup';
```
Comment (ao-alex) [inline src/validations/app.js]: you need to modify package.json to include yup

File: src/i18n/translations/ru.json [hunk 14]
```diff
@@ -1,4 +1,5 @@
 {
+  "is_required": "Это поле обязательно к заполнению",
```
Comment (ao-alex) [inline src/i18n/translations/ru.json]: any particular reason to not have camel case  here?

File: package.json [hunk 15]
```diff
@@ -11,11 +12,13 @@
     "react": "^16.9.0",
     "react-datepicker": "^2.9.6",
     "react-dom": "^16.9.0",
+    "react-hook-form": "^3.23.8",
     "react-i18next": "^10.12.3",
     "react-router": "5.1.2",
     "react-router-dom": "^5.1.2",
     "react-scripts": "3.1.1",
-    "semantic-ui-react": "^0.88.1"
+    "semantic-ui-react": "^0.88.1",
```
Comment (hpilosyan) [inline package.json]: Why do you need semantic ui in this PR?

File: src/validations/app.js [hunk 16]
```diff
@@ -0,0 +1,11 @@
...
+import i18n from '../i18n';
+
+const appSchema = () => Yup.object().shape({
+  username: Yup.string()
+    .required(() => i18n.t('isRequired')),
```
Comment (ao-alex) [inline src/validations/app.js]: use i18n in jsx directly, not here, this should help with update issue
